rpc/RPCTransaction.java

```diff
@@ -238,23 +224,36 @@ private synchronized void clear() {
 
         void add(Response response) {
             responseBuffer.add(response);
-            if (!(response instanceof Response.Ok) || isLastResponse(response.read())) isDone.set(true);
+            if (!(response instanceof Response.Result) || isLastResponse(read(response))) isDone.set(true);
         }
 
         boolean isDone() {
             return isDone.get();
         }
 
         TransactionProto.Transaction.Res take() throws InterruptedException {
-            return responseBuffer.take().read();
+            return read(responseBuffer.take());
         }
 
         TransactionProto.Transaction.Res take(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
             final Response response = responseBuffer.poll(timeout, unit);
-            if (response != null) return response.read();
+            if (response != null) return read(response);
             else throw new TimeoutException();
         }
 
+        TransactionProto.Transaction.Res read(Response response) {
```

Comment:
Why did we take `read` out of `Response`? This method shows the very reason why I put it _in_ `Response` in the first place- effectively a big switch statement with several `instanceof`s. It would be cleaner to take advantage of inheritance, and make it a method inside `Response` that is overloaded in `Result`, `Done.Completed` and `Done.Error`.

Reply:
@lolski suggested that change. I think the reasoning to make the code reads better in the top level `close` method. It would be counter-intuitive if we do `close(new Response.Done.Error(...))` because it's not clear why a normal `close` operation would cause error. Instead we do `close(new Response.Done.Completed(...))` which is more expected.

Reply:
That said, it seems that we should still be able to use inheritance. Let me check.